insights-client wait for inventory and advisor

pholica opened the pull request:
Helper methods for waiting for the system to appear in the inventory and in the advisor.
Those methods should be used only if really needed and exist just because of the design of the services and tooling. It is expected that customer has long running systems and is acceptable that those do not function properly for some time after the registration.

pholica committed:
feat: New method wait_for_inventory

Some insights-client functionality is available only after there's
a record of the system in the inventory, and some tests may require
waiting for such moment, otherwise they would be not able to test
the behaviour of the system once the system appears in the inventory.

pholica committed:
feat: New method wait_for_advisor

Similar as wait_for_inventory, but waiting for advisor.

m-horky commented on pytest_client_tools/insights_client.py:
```diff
@@ -359,3 +360,57 @@ def unregister(self, selinux_context=None):
+        return response.json()["results"][0]["id"]
+
+    def wait_for_inventory(self, *args, **kwargs):
+        return loop_until(lambda: self.inventory_id, *args, **kwargs)
```
I'd like to see a short docstring explaining what happens to `*args, **kwargs` when passed in.

pholica replied:
Docstrings should be resolved now, please close this thread if it makes sense :)

m-horky commented on pytest_client_tools/insights_client.py:
```diff
@@ -359,3 +360,57 @@ def unregister(self, selinux_context=None):
         return self.run("--unregister", selinux_context=selinux_context)
+
+    @property
+    def inventory_id(self):
```
Is there a design reason for this to be a property and not a method? I think it would be nice if the API for 'inventory_id' and 'in_advisor' was the same. Non-blocking note.

pholica replied:
I could change this to a method like `get_inventory_id` I just thought it would be nice to have it as a property for easier use. The property itself is non blocking in the sense that it does not wait for the system to appear in the inventory and returns the current inventory ID (which may not exist and in that case raises exception - thinking about it now, it's not the nicest approach, but exception provides more information than possible plain `None`).

For a moment I was thinking about what you mean by the `in_advisor` API as there's nothing like that and then I've realized that it is there, but just as a local function. Would you prefer making the `in_advisor` a method? And if yes, would you prefer both to be methods `get_inventory_id` and (I don't know) `check_in_advisor` or would properties `inventory_id` and `in_advisor` be better? I really don't have strong preference, just mild inclination to properties in this case.

m-horky replied:
> then I've realized that it is there, but just as a local function
Yes, that's the one I meant.

It is a bit unfortunate one method returns string and the other boolean :/ I can see `check_inventory()` and `check_advisor()` being generic enough for it to not matter.

As I said, it is non-blocking, but it felt note-worthy.

pholica replied:
Ok, as discussed in person, I'll change them to methods `_get_inventory_id` and `_check_advisor`.

Thanks for the feedback, I'm still waking up after the new year and incorrectly interpreted the note about this not blocking this PR :D

pholica replied:
Renamed, marking as resolved.

m-horky commented on pytest_client_tools/insights_client.py:
```diff
@@ -359,3 +360,57 @@ def unregister(self, selinux_context=None):
+    def wait_for_inventory(self, *args, **kwargs):
+        return loop_until(lambda: self.inventory_id, *args, **kwargs)
+
+    def wait_for_advisor(self, *args, **kwargs):
```
Same here for the docstring.

pholica replied:
Docstrings should be resolved now, please close this thread if it makes sense :)

m-horky reviewed (changes requested):
I like the approach. I left some notes that could improve maintainability over time.

pholica committed:
feat: Add exception handling to loop_until

The exception needs to be explicitly saved because of flake8
which incorrectly thinks the `exception` variable is unused
with plain use in the except statement like:
```
exception = None
try:
    ...
except Exception as exception:
    pass
print(exception)
```

m-horky reviewed (changes requested):
Overall I like the approach. Since `requests` doesn't have any timeout set by default, I think we should set _some_ value in there, something that would make sense in the context of default test timeout (anything between 10 and 60 seconds would make sense to me).

pkoprda commented on pytest_client_tools/insights_client.py:
```diff
@@ -359,3 +360,128 @@ def unregister(self, selinux_context=None):
+        """
+        inventory_url = (
+            f"https://{self._get_services_api_host()}"
+            f"/api/inventory/v1/hosts?insights_id={self.uuid}"
```
Wouldn't be better to use `host_exists` endpoint? It doesn't overload the HBI database. `host_exists` is also called when various commands are used by insights-client. Non-blocking

pholica replied:
Does `host_exists` provide inventory id? If yes, It can be surely used. I just don't know the API at all :/

pkoprda replied:
Yes, `host_exists` finds one host by insights_id, if it exists. See an example of the usage in insights-core repository: https://github.com/RedHatInsights/insights-core/blob/b22c6c0a833a74a8936f57897519e1ad65dc9e15/insights/client/connection.py#L750

pholica replied:
Ok, changing it to this then. Let's see if it works.

pkoprda reviewed (approved):
LGTM

m-horky reviewed (approved):
ACK
